## 1 - Cargar datos

In [316]:
import pandas as pd

In [317]:
relative_path = "../datasets/json_data/customers.json"
df_json = pd.read_json(relative_path, lines = True)

In [318]:
df_json.head(5)

fraudulent                                           customer  \
0       False  {'customerEmail': 'josephhoward@yahoo.com', 'c...   
1        True  {'customerEmail': 'evansjeffery@yahoo.com', 'c...   
2       False  {'customerEmail': 'andersonwilliam@yahoo.com',...   
3       False  {'customerEmail': 'rubenjuarez@yahoo.com', 'cu...   
4        True  {'customerEmail': 'uchen@malone.com', 'custome...   

                                              orders  \
0  [{'orderId': 'vjbdvd', 'orderAmount': 18, 'ord...   
1  [{'orderId': 'nlghpa', 'orderAmount': 45, 'ord...   
2  [{'orderId': 'yk34y2', 'orderAmount': 33, 'ord...   
3  [{'orderId': 'fbz9ep', 'orderAmount': 34, 'ord...   
4  [{'orderId': '56h7iw', 'orderAmount': 71, 'ord...   

                                      paymentMethods  \
0  [{'paymentMethodId': 'wt07xm68b', 'paymentMeth...   
1  [{'paymentMethodId': 'y3xp697jx', 'paymentMeth...   
2  [{'paymentMethodId': '8pneoi03z', 'paymentMeth...   
3  [{'paymentMethodId': 'pdxjdwui4', 'paymentMeth...   
4  [{'paymentMethodId': 'w1i1zq3rg', 'paymentMeth...   

                                        transactions  
0  [{'transactionId': 'a9lcj51r', 'orderId': 'vjb...  
1  [{'transactionId': '5mi94sfw', 'orderId': 'nlg...  
2  [{'transactionId': 'q3lyvbza', 'orderId': 'yk3...  
3  [{'transactionId': 'vx4cjc27', 'orderId': 'fbz...  
4  [{'transactionId': 'q8j3dgni', 'orderId': '56h...

In [319]:
df_json.tail(5)

fraudulent                                           customer  \
163        True  {'customerEmail': 'mitchellvickie@brewer-jones...   
164        True  {'customerEmail': 'sbrown@hughes.biz', 'custom...   
165        True  {'customerEmail': 'johnlowery@gmail.com', 'cus...   
166        True  {'customerEmail': 'ethompson@jackson-sanders.c...   
167       False  {'customerEmail': 'dawn05@tucker-brown.com', '...   

                                                orders  \
163  [{'orderId': '6086rd', 'orderAmount': 31, 'ord...   
164  [{'orderId': '57yhdo', 'orderAmount': 30, 'ord...   
165  [{'orderId': 'vcyuse', 'orderAmount': 35, 'ord...   
166  [{'orderId': '9s9ne9', 'orderAmount': 49, 'ord...   
167  [{'orderId': 'ow1cih', 'orderAmount': 23, 'ord...   

                                        paymentMethods  \
163  [{'paymentMethodId': 'x644o186l', 'paymentMeth...   
164  [{'paymentMethodId': 'wsy5c41nf', 'paymentMeth...   
165  [{'paymentMethodId': 'j3foo2l16', 'paymentMeth...   
166  [{'paymentMethodId': 'b3jl38c9g', 'paymentMeth...   
167  [{'paymentMethodId': 'f0xnu5om9', 'paymentMeth...   

                                          transactions  
163  [{'transactionId': 'ryxa1mzn', 'orderId': '608...  
164  [{'transactionId': '293g6767', 'orderId': '57y...  
165  [{'transactionId': 'qio8vvq7', 'orderId': 'vcy...  
166  [{'transactionId': '1uusmhj9', 'orderId': '9s9...  
167  [{'transactionId': '6incxb5x', 'orderId': 'ow1...

## 2 - Desanidar columnas

Como tenemos ciertas columnas con su información en formato _json_, vamos a convertir las columnas, por separado, en un _dataFrame_.

##### Customer

In [320]:
from pandas import json_normalize

In [321]:
df_json["customer"]

0      {'customerEmail': 'josephhoward@yahoo.com', 'c...
1      {'customerEmail': 'evansjeffery@yahoo.com', 'c...
2      {'customerEmail': 'andersonwilliam@yahoo.com',...
3      {'customerEmail': 'rubenjuarez@yahoo.com', 'cu...
4      {'customerEmail': 'uchen@malone.com', 'custome...
                             ...                        
163    {'customerEmail': 'mitchellvickie@brewer-jones...
164    {'customerEmail': 'sbrown@hughes.biz', 'custom...
165    {'customerEmail': 'johnlowery@gmail.com', 'cus...
166    {'customerEmail': 'ethompson@jackson-sanders.c...
167    {'customerEmail': 'dawn05@tucker-brown.com', '...
Name: customer, Length: 168, dtype: object

In [322]:
customers_df = json_normalize(df_json["customer"])

In [323]:
customers_df.head(5)

customerEmail       customerPhone        customerDevice  \
0     josephhoward@yahoo.com        400-108-5415  yyeiaxpltf82440jnb3v   
1     evansjeffery@yahoo.com      1-788-091-7546  r0jpm7xaeqqa3kr6mzum   
2  andersonwilliam@yahoo.com        024.420.0375  4m7h5ipl1shyavt6vv2r   
3      rubenjuarez@yahoo.com  670.664.8168x94985  slovx60t0i558may4ks0   
4           uchen@malone.com      1-981-877-0870  j0pd24k5h8dl2fqu0cz4   

                         customerIPAddress  \
0                             8.129.104.40   
1                          219.173.211.202   
2  67b7:3db8:67e0:3bea:b9d0:90c1:2b60:b9f0   
3  95de:8565:5a66:792c:26e0:6cfb:7d87:11af   
4                           196.89.235.192   

                              customerBillingAddress  
0            5493 Jones Islands\nBrownside, CA 51896  
1  356 Elizabeth Inlet Suite 120\nPort Joshuabury...  
2     8478 Sean Ridges Apt. 441\nDavisberg, PR 72250  
3  7769 Elizabeth Bridge Apt. 343\nNortonstad, FM...  
4  148 Russell Lodge Apt. 445\nPort Jenniferside,...

##### Orders

In [324]:
df_json["orders"]

0      [{'orderId': 'vjbdvd', 'orderAmount': 18, 'ord...
1      [{'orderId': 'nlghpa', 'orderAmount': 45, 'ord...
2      [{'orderId': 'yk34y2', 'orderAmount': 33, 'ord...
3      [{'orderId': 'fbz9ep', 'orderAmount': 34, 'ord...
4      [{'orderId': '56h7iw', 'orderAmount': 71, 'ord...
                             ...                        
163    [{'orderId': '6086rd', 'orderAmount': 31, 'ord...
164    [{'orderId': '57yhdo', 'orderAmount': 30, 'ord...
165    [{'orderId': 'vcyuse', 'orderAmount': 35, 'ord...
166    [{'orderId': '9s9ne9', 'orderAmount': 49, 'ord...
167    [{'orderId': 'ow1cih', 'orderAmount': 23, 'ord...
Name: orders, Length: 168, dtype: object

In [325]:
orders_nested = pd.DataFrame([md for md in df_json["orders"]])

order_list = []
for index, row in orders_nested.iterrows():
    for order in row:
        if order != None:
            order_list.append(order)

orders_df = pd.DataFrame(order_list)

In [326]:
orders_df.head(5)

orderId  orderAmount orderState  \
0  vjbdvd           18    pending   
1  yp6x27           26  fulfilled   
2  nlghpa           45  fulfilled   
3  uw0eeb           23  fulfilled   
4  bn44oh           43  fulfilled   

                                orderShippingAddress  
0            5493 Jones Islands\nBrownside, CA 51896  
1            5493 Jones Islands\nBrownside, CA 51896  
2      898 Henry Ports\nNew Keithview, CA 95893-2497  
3  356 Elizabeth Inlet Suite 120\nPort Joshuabury...  
4         5093 Bryan Forks\nJoshuaton, FM 01565-9801

##### Transactions

In [327]:
df_json["transactions"]

0      [{'transactionId': 'a9lcj51r', 'orderId': 'vjb...
1      [{'transactionId': '5mi94sfw', 'orderId': 'nlg...
2      [{'transactionId': 'q3lyvbza', 'orderId': 'yk3...
3      [{'transactionId': 'vx4cjc27', 'orderId': 'fbz...
4      [{'transactionId': 'q8j3dgni', 'orderId': '56h...
                             ...                        
163    [{'transactionId': 'ryxa1mzn', 'orderId': '608...
164    [{'transactionId': '293g6767', 'orderId': '57y...
165    [{'transactionId': 'qio8vvq7', 'orderId': 'vcy...
166    [{'transactionId': '1uusmhj9', 'orderId': '9s9...
167    [{'transactionId': '6incxb5x', 'orderId': 'ow1...
Name: transactions, Length: 168, dtype: object

In [328]:
transactions_nested = pd.DataFrame([md for md in df_json["transactions"]])

transactions_list = []
for index, row in transactions_nested.iterrows():
    for transaction in row:
        if transaction != None:
            transactions_list.append(transaction)

transactions_df = pd.DataFrame(transactions_list)

In [329]:
transactions_df.head(5)

transactionId orderId paymentMethodId  transactionAmount  transactionFailed
0      a9lcj51r  vjbdvd       wt07xm68b                 18              False
1      y4wcv03i  yp6x27       wt07xm68b                 26              False
2      5mi94sfw  nlghpa       41ug157xz                 45              False
3      br8ba1nu  uw0eeb       41ug157xz                 23              False
4      a33145ss  bn44oh       y3xp697jx                 43               True

##### Payment

In [330]:
df_json["paymentMethods"]

0      [{'paymentMethodId': 'wt07xm68b', 'paymentMeth...
1      [{'paymentMethodId': 'y3xp697jx', 'paymentMeth...
2      [{'paymentMethodId': '8pneoi03z', 'paymentMeth...
3      [{'paymentMethodId': 'pdxjdwui4', 'paymentMeth...
4      [{'paymentMethodId': 'w1i1zq3rg', 'paymentMeth...
                             ...                        
163    [{'paymentMethodId': 'x644o186l', 'paymentMeth...
164    [{'paymentMethodId': 'wsy5c41nf', 'paymentMeth...
165    [{'paymentMethodId': 'j3foo2l16', 'paymentMeth...
166    [{'paymentMethodId': 'b3jl38c9g', 'paymentMeth...
167    [{'paymentMethodId': 'f0xnu5om9', 'paymentMeth...
Name: paymentMethods, Length: 168, dtype: object

In [331]:
payment_nested = pd.DataFrame([md for md in df_json["paymentMethods"]])

payment_list = []
for index, row in payment_nested.iterrows():
    for payment in row:
        if payment != None:
            payment_list.append(payment)

payments_df = pd.DataFrame(payment_list)

In [332]:
payments_df.head(5)

paymentMethodId  paymentMethodRegistrationFailure paymentMethodType  \
0       wt07xm68b                              True              card   
1       y3xp697jx                              True           bitcoin   
2       6krszxc05                             False              card   
3       5z1szj2he                             False              card   
4       m52tx8e1s                             False              card   

         paymentMethodProvider     paymentMethodIssuer  
0                 JCB 16 digit    Citizens First Banks  
1                VISA 16 digit            Solace Banks  
2                VISA 16 digit          Vertex Bancorp  
3  Diners Club / Carte Blanche  His Majesty Bank Corp.  
4                   Mastercard          Vertex Bancorp

##### Fraudulent

In [333]:
fraudulent_df = df_json["fraudulent"]

In [334]:
fraudulent_df.head(5)

0    False
1     True
2    False
3    False
4     True
Name: fraudulent, dtype: bool

## 3 - Mergear dataframes

Ahora vamos a unir todos los dataframes de las columnas que convertimos.

In [335]:
data = pd.concat([fraudulent_df, payments_df, transactions_df, orders_df, customers_df], axis = 1)

In [336]:
data.head(5)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   
3      False       5z1szj2he                            False   
4       True       m52tx8e1s                            False   

  paymentMethodType        paymentMethodProvider     paymentMethodIssuer  \
0              card                 JCB 16 digit    Citizens First Banks   
1           bitcoin                VISA 16 digit            Solace Banks   
2              card                VISA 16 digit          Vertex Bancorp   
3              card  Diners Club / Carte Blanche  His Majesty Bank Corp.   
4              card                   Mastercard          Vertex Bancorp   

  transactionId orderId paymentMethodId  transactionAmount  transactionFailed  \
0      a9lcj51r  vjbdvd       wt07xm68b                 18              False   
1      y4wcv03i  yp6x27       wt07xm68b                 26              False   
2      5mi94sfw  nlghpa       41ug157xz                 45              False   
3      br8ba1nu  uw0eeb       41ug157xz                 23              False   
4      a33145ss  bn44oh       y3xp697jx                 43               True   

  orderId  orderAmount orderState  \
0  vjbdvd         18.0    pending   
1  yp6x27         26.0  fulfilled   
2  nlghpa         45.0  fulfilled   
3  uw0eeb         23.0  fulfilled   
4  bn44oh         43.0  fulfilled   

                                orderShippingAddress  \
0            5493 Jones Islands\nBrownside, CA 51896   
1            5493 Jones Islands\nBrownside, CA 51896   
2      898 Henry Ports\nNew Keithview, CA 95893-2497   
3  356 Elizabeth Inlet Suite 120\nPort Joshuabury...   
4         5093 Bryan Forks\nJoshuaton, FM 01565-9801   

               customerEmail       customerPhone        customerDevice  \
0     josephhoward@yahoo.com        400-108-5415  yyeiaxpltf82440jnb3v   
1     evansjeffery@yahoo.com      1-788-091-7546  r0jpm7xaeqqa3kr6mzum   
2  andersonwilliam@yahoo.com        024.420.0375  4m7h5ipl1shyavt6vv2r   
3      rubenjuarez@yahoo.com  670.664.8168x94985  slovx60t0i558may4ks0   
4           uchen@malone.com      1-981-877-0870  j0pd24k5h8dl2fqu0cz4   

                         customerIPAddress  \
0                             8.129.104.40   
1                          219.173.211.202   
2  67b7:3db8:67e0:3bea:b9d0:90c1:2b60:b9f0   
3  95de:8565:5a66:792c:26e0:6cfb:7d87:11af   
4                           196.89.235.192   

                              customerBillingAddress  
0            5493 Jones Islands\nBrownside, CA 51896  
1  356 Elizabeth Inlet Suite 120\nPort Joshuabury...  
2     8478 Sean Ridges Apt. 441\nDavisberg, PR 72250  
3  7769 Elizabeth Bridge Apt. 343\nNortonstad, FM...  
4  148 Russell Lodge Apt. 445\nPort Jenniferside,...

## 4 - Transformación de columnas

### 4.1 - Columnas duplicadas

In [337]:
from funpymodeling import status

In [338]:
status(data)

variable  q_nan     p_nan  q_zeros   p_zeros  \
0                         fraudulent    455  0.730337      107  0.171750   
1                    paymentMethodId    281  0.451043        0  0.000000   
2   paymentMethodRegistrationFailure    281  0.451043      302  0.484751   
3                  paymentMethodType    281  0.451043        0  0.000000   
4              paymentMethodProvider    281  0.451043        0  0.000000   
5                paymentMethodIssuer    281  0.451043        0  0.000000   
6                      transactionId      0  0.000000        0  0.000000   
7                            orderId      0  0.000000        0  0.000000   
8                    paymentMethodId      0  0.000000        0  0.000000   
9                  transactionAmount      0  0.000000        0  0.000000   
10                 transactionFailed      0  0.000000      455  0.730337   
11                           orderId    145  0.232745        0  0.000000   
12                       orderAmount    145  0.232745        0  0.000000   
13                        orderState    145  0.232745        0  0.000000   
14              orderShippingAddress    145  0.232745        0  0.000000   
15                     customerEmail    455  0.730337        0  0.000000   
16                     customerPhone    455  0.730337        0  0.000000   
17                    customerDevice    455  0.730337        0  0.000000   
18                 customerIPAddress    455  0.730337        0  0.000000   
19            customerBillingAddress    455  0.730337        0  0.000000   

    unique     type  
0        2   object  
1      342   object  
2        2   object  
3        4   object  
4       10   object  
5       20   object  
6      623   object  
7      478   object  
8      237   object  
9       67    int64  
10       2     bool  
11     478   object  
12      67  float64  
13       3   object  
14     274   object  
15     161   object  
16     168   object  
17     168   object  
18     165   object  
19     166   object

In [339]:
data = data.loc[:,~data.columns.duplicated()].copy()

In [340]:
status(data)

variable  q_nan     p_nan  q_zeros   p_zeros  \
0                         fraudulent    455  0.730337      107  0.171750   
1                    paymentMethodId    281  0.451043        0  0.000000   
2   paymentMethodRegistrationFailure    281  0.451043      302  0.484751   
3                  paymentMethodType    281  0.451043        0  0.000000   
4              paymentMethodProvider    281  0.451043        0  0.000000   
5                paymentMethodIssuer    281  0.451043        0  0.000000   
6                      transactionId      0  0.000000        0  0.000000   
7                            orderId      0  0.000000        0  0.000000   
8                  transactionAmount      0  0.000000        0  0.000000   
9                  transactionFailed      0  0.000000      455  0.730337   
10                       orderAmount    145  0.232745        0  0.000000   
11                        orderState    145  0.232745        0  0.000000   
12              orderShippingAddress    145  0.232745        0  0.000000   
13                     customerEmail    455  0.730337        0  0.000000   
14                     customerPhone    455  0.730337        0  0.000000   
15                    customerDevice    455  0.730337        0  0.000000   
16                 customerIPAddress    455  0.730337        0  0.000000   
17            customerBillingAddress    455  0.730337        0  0.000000   

    unique     type  
0        2   object  
1      342   object  
2        2   object  
3        4   object  
4       10   object  
5       20   object  
6      623   object  
7      478   object  
8       67    int64  
9        2     bool  
10      67  float64  
11       3   object  
12     274   object  
13     161   object  
14     168   object  
15     168   object  
16     165   object  
17     166   object

In [341]:
data.head(3)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   

  paymentMethodType paymentMethodProvider   paymentMethodIssuer transactionId  \
0              card          JCB 16 digit  Citizens First Banks      a9lcj51r   
1           bitcoin         VISA 16 digit          Solace Banks      y4wcv03i   
2              card         VISA 16 digit        Vertex Bancorp      5mi94sfw   

  orderId  transactionAmount  transactionFailed  orderAmount orderState  \
0  vjbdvd                 18              False         18.0    pending   
1  yp6x27                 26              False         26.0  fulfilled   
2  nlghpa                 45              False         45.0  fulfilled   

                            orderShippingAddress              customerEmail  \
0        5493 Jones Islands\nBrownside, CA 51896     josephhoward@yahoo.com   
1        5493 Jones Islands\nBrownside, CA 51896     evansjeffery@yahoo.com   
2  898 Henry Ports\nNew Keithview, CA 95893-2497  andersonwilliam@yahoo.com   

    customerPhone        customerDevice  \
0    400-108-5415  yyeiaxpltf82440jnb3v   
1  1-788-091-7546  r0jpm7xaeqqa3kr6mzum   
2    024.420.0375  4m7h5ipl1shyavt6vv2r   

                         customerIPAddress  \
0                             8.129.104.40   
1                          219.173.211.202   
2  67b7:3db8:67e0:3bea:b9d0:90c1:2b60:b9f0   

                              customerBillingAddress  
0            5493 Jones Islands\nBrownside, CA 51896  
1  356 Elizabeth Inlet Suite 120\nPort Joshuabury...  
2     8478 Sean Ridges Apt. 441\nDavisberg, PR 72250

### 4.2 - Customer Email

Los tipos de mails más comúnes dentro de los clientes son, `hotmail`, `gmail` y `yahoo`.

Como muchas veces, ciertos clientes con intenciones maliciosas usan otro tipo de mail, como del tipo `@1jcfcxs7`, los podemos clasificar como `weird`.

In [342]:
email_domains = []
mails = []
popular_providers = ["yahoo", "gmail", "hotmail"]

for email in data["customerEmail"]:
    try:
        aux = email.split("@")[1]
        domain = aux.split(".")[1]
        mail = aux.split(".")[0]

        if mail in popular_providers:
            mails.append(mail)
        else:
            mails.append("other")
        email_domains.append(domain)
    except:
        email_domains.append("weird")
        mails.append("weird")

In [343]:
data["emailDomain"] = email_domains
data["emailProvider"] = mails

In [344]:
data.drop(["customerEmail"],inplace=True, axis=1)

In [345]:
data.head(5)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   
3      False       5z1szj2he                            False   
4       True       m52tx8e1s                            False   

  paymentMethodType        paymentMethodProvider     paymentMethodIssuer  \
0              card                 JCB 16 digit    Citizens First Banks   
1           bitcoin                VISA 16 digit            Solace Banks   
2              card                VISA 16 digit          Vertex Bancorp   
3              card  Diners Club / Carte Blanche  His Majesty Bank Corp.   
4              card                   Mastercard          Vertex Bancorp   

  transactionId orderId  transactionAmount  transactionFailed  orderAmount  \
0      a9lcj51r  vjbdvd                 18              False         18.0   
1      y4wcv03i  yp6x27                 26              False         26.0   
2      5mi94sfw  nlghpa                 45              False         45.0   
3      br8ba1nu  uw0eeb                 23              False         23.0   
4      a33145ss  bn44oh                 43               True         43.0   

  orderState                               orderShippingAddress  \
0    pending            5493 Jones Islands\nBrownside, CA 51896   
1  fulfilled            5493 Jones Islands\nBrownside, CA 51896   
2  fulfilled      898 Henry Ports\nNew Keithview, CA 95893-2497   
3  fulfilled  356 Elizabeth Inlet Suite 120\nPort Joshuabury...   
4  fulfilled         5093 Bryan Forks\nJoshuaton, FM 01565-9801   

        customerPhone        customerDevice  \
0        400-108-5415  yyeiaxpltf82440jnb3v   
1      1-788-091-7546  r0jpm7xaeqqa3kr6mzum   
2        024.420.0375  4m7h5ipl1shyavt6vv2r   
3  670.664.8168x94985  slovx60t0i558may4ks0   
4      1-981-877-0870  j0pd24k5h8dl2fqu0cz4   

                         customerIPAddress  \
0                             8.129.104.40   
1                          219.173.211.202   
2  67b7:3db8:67e0:3bea:b9d0:90c1:2b60:b9f0   
3  95de:8565:5a66:792c:26e0:6cfb:7d87:11af   
4                           196.89.235.192   

                              customerBillingAddress emailDomain emailProvider  
0            5493 Jones Islands\nBrownside, CA 51896         com         yahoo  
1  356 Elizabeth Inlet Suite 120\nPort Joshuabury...         com         yahoo  
2     8478 Sean Ridges Apt. 441\nDavisberg, PR 72250         com         yahoo  
3  7769 Elizabeth Bridge Apt. 343\nNortonstad, FM...         com         yahoo  
4  148 Russell Lodge Apt. 445\nPort Jenniferside,...         com         other

### 4.3 - CustomerIPAdress

In [346]:
from funpymodeling import freq_tbl

freq_tbl(data["customerIPAddress"])

customerIPAddress  frequency  percentage  \
0                              45.203.99.249          4    0.006421   
1                               8.129.104.40          1    0.001605   
2                             11.165.157.167          1    0.001605   
3                               181.88.44.76          1    0.001605   
4                             242.102.134.32          1    0.001605   
..                                       ...        ...         ...   
160  13ff:5a93:bc34:e190:5e4b:ac80:2e2d:6217          1    0.001605   
161                          115.137.171.103          1    0.001605   
162    710e:d9d9:1ef0:cdc:ed53:10d:2c7e:667e          1    0.001605   
163                             82.3.174.184          1    0.001605   
164                            132.132.95.33          1    0.001605   

     cumulative_perc  
0           0.023810  
1           0.029762  
2           0.035714  
3           0.041667  
4           0.047619  
..               ...  
160         0.976190  
161         0.982143  
162         0.988095  
163         0.994048  
164         1.000000  

[165 rows x 4 columns]

In [347]:
len("255.255.255.255") #Mayor cantidad de caracteres numéricos que puede tener una subnet mask

15

Debido a que hay IPs más extensas con más de 15 carácteres y además, esos caracteres no son solo números sino que también letras.

Por eso mismo, creamos un nueva columna `CustomerIPAddressSimplified` con 2 valores: `only_numeric_digits` y `digits_and_letters` para cada tipo de dirección IP según su longitud y tipo.

In [348]:
IP_adresses = []

for adress in data["customerIPAddress"]:
    aux_adress = str(adress)

    if len(aux_adress) > 15:
        IP_adresses.append("digits_and_letters")
    else:
        IP_adresses.append("only_numeric_digits")

In [349]:
data["customerIPAddressSimplified"] = IP_adresses

In [350]:
data.drop(["customerIPAddress"],axis=1,inplace=True)

In [351]:
data.head(5)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   
3      False       5z1szj2he                            False   
4       True       m52tx8e1s                            False   

  paymentMethodType        paymentMethodProvider     paymentMethodIssuer  \
0              card                 JCB 16 digit    Citizens First Banks   
1           bitcoin                VISA 16 digit            Solace Banks   
2              card                VISA 16 digit          Vertex Bancorp   
3              card  Diners Club / Carte Blanche  His Majesty Bank Corp.   
4              card                   Mastercard          Vertex Bancorp   

  transactionId orderId  transactionAmount  transactionFailed  orderAmount  \
0      a9lcj51r  vjbdvd                 18              False         18.0   
1      y4wcv03i  yp6x27                 26              False         26.0   
2      5mi94sfw  nlghpa                 45              False         45.0   
3      br8ba1nu  uw0eeb                 23              False         23.0   
4      a33145ss  bn44oh                 43               True         43.0   

  orderState                               orderShippingAddress  \
0    pending            5493 Jones Islands\nBrownside, CA 51896   
1  fulfilled            5493 Jones Islands\nBrownside, CA 51896   
2  fulfilled      898 Henry Ports\nNew Keithview, CA 95893-2497   
3  fulfilled  356 Elizabeth Inlet Suite 120\nPort Joshuabury...   
4  fulfilled         5093 Bryan Forks\nJoshuaton, FM 01565-9801   

        customerPhone        customerDevice  \
0        400-108-5415  yyeiaxpltf82440jnb3v   
1      1-788-091-7546  r0jpm7xaeqqa3kr6mzum   
2        024.420.0375  4m7h5ipl1shyavt6vv2r   
3  670.664.8168x94985  slovx60t0i558may4ks0   
4      1-981-877-0870  j0pd24k5h8dl2fqu0cz4   

                              customerBillingAddress emailDomain  \
0            5493 Jones Islands\nBrownside, CA 51896         com   
1  356 Elizabeth Inlet Suite 120\nPort Joshuabury...         com   
2     8478 Sean Ridges Apt. 441\nDavisberg, PR 72250         com   
3  7769 Elizabeth Bridge Apt. 343\nNortonstad, FM...         com   
4  148 Russell Lodge Apt. 445\nPort Jenniferside,...         com   

  emailProvider customerIPAddressSimplified  
0         yahoo         only_numeric_digits  
1         yahoo         only_numeric_digits  
2         yahoo          digits_and_letters  
3         yahoo          digits_and_letters  
4         other         only_numeric_digits

### 4.4 - customerBillingAddress and orderShippingAdress

In [352]:
data["customerBillingAddress"].head(5)

0              5493 Jones Islands\nBrownside, CA 51896
1    356 Elizabeth Inlet Suite 120\nPort Joshuabury...
2       8478 Sean Ridges Apt. 441\nDavisberg, PR 72250
3    7769 Elizabeth Bridge Apt. 343\nNortonstad, FM...
4    148 Russell Lodge Apt. 445\nPort Jenniferside,...
Name: customerBillingAddress, dtype: object

In [353]:
data["orderShippingAddress"].head(5)

0              5493 Jones Islands\nBrownside, CA 51896
1              5493 Jones Islands\nBrownside, CA 51896
2        898 Henry Ports\nNew Keithview, CA 95893-2497
3    356 Elizabeth Inlet Suite 120\nPort Joshuabury...
4           5093 Bryan Forks\nJoshuaton, FM 01565-9801
Name: orderShippingAddress, dtype: object

Vamos a buscar quedarnos unicamente con la ciudad, _Por ejemplo_: Del string `898 Henry Ports\nNew Keithview, CA 95893-2497` nos quedamos con `CA`.

In [354]:
def extract_city_name(name_column):
    city_list = []

    for adress in data[name_column]:
        try:
            aux = adress.split("\n")[1]
            city_with_number = aux.split(", ")[1]
            city = city_with_number.split(" ")[0]

            city_list.append(city)

        except:
            city_list.append("unknown")
    
    return city_list

In [355]:
data["customerBillingAddress"] = extract_city_name("customerBillingAddress")
data["orderShippingAddress"] = extract_city_name("orderShippingAddress")

In [356]:
data.head(5)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   
3      False       5z1szj2he                            False   
4       True       m52tx8e1s                            False   

  paymentMethodType        paymentMethodProvider     paymentMethodIssuer  \
0              card                 JCB 16 digit    Citizens First Banks   
1           bitcoin                VISA 16 digit            Solace Banks   
2              card                VISA 16 digit          Vertex Bancorp   
3              card  Diners Club / Carte Blanche  His Majesty Bank Corp.   
4              card                   Mastercard          Vertex Bancorp   

  transactionId orderId  transactionAmount  transactionFailed  orderAmount  \
0      a9lcj51r  vjbdvd                 18              False         18.0   
1      y4wcv03i  yp6x27                 26              False         26.0   
2      5mi94sfw  nlghpa                 45              False         45.0   
3      br8ba1nu  uw0eeb                 23              False         23.0   
4      a33145ss  bn44oh                 43               True         43.0   

  orderState orderShippingAddress       customerPhone        customerDevice  \
0    pending                   CA        400-108-5415  yyeiaxpltf82440jnb3v   
1  fulfilled                   CA      1-788-091-7546  r0jpm7xaeqqa3kr6mzum   
2  fulfilled                   CA        024.420.0375  4m7h5ipl1shyavt6vv2r   
3  fulfilled                   NM  670.664.8168x94985  slovx60t0i558may4ks0   
4  fulfilled                   FM      1-981-877-0870  j0pd24k5h8dl2fqu0cz4   

  customerBillingAddress emailDomain emailProvider customerIPAddressSimplified  
0                     CA         com         yahoo         only_numeric_digits  
1                     NM         com         yahoo         only_numeric_digits  
2                     PR         com         yahoo          digits_and_letters  
3                     FM         com         yahoo          digits_and_letters  
4                     TN         com         other         only_numeric_digits

Ahora vamos a evaluar si las 2 ciudades que encontramos en `customerBillingAddress` y `orderShippingAddress` son las mismas, ya que, en el caso de no ser la misma, podría indicar la posibilidad de que sea fraude.

De esa manera, creamos una nueva columna `sameCity`, donde va a tener como valores, **yes**, **no** y **unknown**

In [357]:
same_city = []

for index in range(len(data)):
    billing_address = data["customerBillingAddress"][index]
    order_address = data["orderShippingAddress"][index]

    response = "unknown"
    if billing_address != response and order_address != response:
        if billing_address == order_address:
            response = "yes"
        else:
            response = "no"

    same_city.append(response)

In [358]:
data["sameCity"] = same_city

In [359]:
data.drop(["customerBillingAddress","orderShippingAddress"],axis=1,inplace=True)

In [360]:
data.head(5)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   
3      False       5z1szj2he                            False   
4       True       m52tx8e1s                            False   

  paymentMethodType        paymentMethodProvider     paymentMethodIssuer  \
0              card                 JCB 16 digit    Citizens First Banks   
1           bitcoin                VISA 16 digit            Solace Banks   
2              card                VISA 16 digit          Vertex Bancorp   
3              card  Diners Club / Carte Blanche  His Majesty Bank Corp.   
4              card                   Mastercard          Vertex Bancorp   

  transactionId orderId  transactionAmount  transactionFailed  orderAmount  \
0      a9lcj51r  vjbdvd                 18              False         18.0   
1      y4wcv03i  yp6x27                 26              False         26.0   
2      5mi94sfw  nlghpa                 45              False         45.0   
3      br8ba1nu  uw0eeb                 23              False         23.0   
4      a33145ss  bn44oh                 43               True         43.0   

  orderState       customerPhone        customerDevice emailDomain  \
0    pending        400-108-5415  yyeiaxpltf82440jnb3v         com   
1  fulfilled      1-788-091-7546  r0jpm7xaeqqa3kr6mzum         com   
2  fulfilled        024.420.0375  4m7h5ipl1shyavt6vv2r         com   
3  fulfilled  670.664.8168x94985  slovx60t0i558may4ks0         com   
4  fulfilled      1-981-877-0870  j0pd24k5h8dl2fqu0cz4         com   

  emailProvider customerIPAddressSimplified sameCity  
0         yahoo         only_numeric_digits      yes  
1         yahoo         only_numeric_digits       no  
2         yahoo          digits_and_letters       no  
3         yahoo          digits_and_letters       no  
4         other         only_numeric_digits       no

### 4.5 - customerPhone

El número de teléfono se podría considerar un identificador único.
Lo que se podría obtener es el código de área, es decir, podríamos extraer el primer número para saber de que país es, y en algunos casos también de que provincia es.

Como en este caso, la manera en la que se registra el dato en `customerPhone` es muy variada y no podemos encontrar un patrón, elegimos eliminar la columna directamente.

In [361]:
data["customerPhone"].head(5)

0          400-108-5415
1        1-788-091-7546
2          024.420.0375
3    670.664.8168x94985
4        1-981-877-0870
Name: customerPhone, dtype: object

In [362]:
data.drop(["customerPhone"],axis=1,inplace=True)

In [363]:
data.head(5)

fraudulent paymentMethodId paymentMethodRegistrationFailure  \
0      False       wt07xm68b                             True   
1       True       y3xp697jx                             True   
2      False       6krszxc05                            False   
3      False       5z1szj2he                            False   
4       True       m52tx8e1s                            False   

  paymentMethodType        paymentMethodProvider     paymentMethodIssuer  \
0              card                 JCB 16 digit    Citizens First Banks   
1           bitcoin                VISA 16 digit            Solace Banks   
2              card                VISA 16 digit          Vertex Bancorp   
3              card  Diners Club / Carte Blanche  His Majesty Bank Corp.   
4              card                   Mastercard          Vertex Bancorp   

  transactionId orderId  transactionAmount  transactionFailed  orderAmount  \
0      a9lcj51r  vjbdvd                 18              False         18.0   
1      y4wcv03i  yp6x27                 26              False         26.0   
2      5mi94sfw  nlghpa                 45              False         45.0   
3      br8ba1nu  uw0eeb                 23              False         23.0   
4      a33145ss  bn44oh                 43               True         43.0   

  orderState        customerDevice emailDomain emailProvider  \
0    pending  yyeiaxpltf82440jnb3v         com         yahoo   
1  fulfilled  r0jpm7xaeqqa3kr6mzum         com         yahoo   
2  fulfilled  4m7h5ipl1shyavt6vv2r         com         yahoo   
3  fulfilled  slovx60t0i558may4ks0         com         yahoo   
4  fulfilled  j0pd24k5h8dl2fqu0cz4         com         other   

  customerIPAddressSimplified sameCity  
0         only_numeric_digits      yes  
1         only_numeric_digits       no  
2          digits_and_letters       no  
3          digits_and_letters       no  
4         only_numeric_digits       no

### 4.6- Identificadores únicos

Podemos encontrar identificadores únicos en las siguientes columnas: `paymentMethodId`, `transactionId`, `orderId` y `customerDevice`.

Debido a que buscamos eliminar estos identificadores en estos tipos de productos de prevención de fraude, eliminamos estas columnas

In [364]:
data.drop(["paymentMethodId", "transactionId", "orderId", "customerDevice"],axis=1,inplace=True)

In [365]:
data.head(5)

fraudulent paymentMethodRegistrationFailure paymentMethodType  \
0      False                             True              card   
1       True                             True           bitcoin   
2      False                            False              card   
3      False                            False              card   
4       True                            False              card   

         paymentMethodProvider     paymentMethodIssuer  transactionAmount  \
0                 JCB 16 digit    Citizens First Banks                 18   
1                VISA 16 digit            Solace Banks                 26   
2                VISA 16 digit          Vertex Bancorp                 45   
3  Diners Club / Carte Blanche  His Majesty Bank Corp.                 23   
4                   Mastercard          Vertex Bancorp                 43   

   transactionFailed  orderAmount orderState emailDomain emailProvider  \
0              False         18.0    pending         com         yahoo   
1              False         26.0  fulfilled         com         yahoo   
2              False         45.0  fulfilled         com         yahoo   
3              False         23.0  fulfilled         com         yahoo   
4               True         43.0  fulfilled         com         other   

  customerIPAddressSimplified sameCity  
0         only_numeric_digits      yes  
1         only_numeric_digits       no  
2          digits_and_letters       no  
3          digits_and_letters       no  
4         only_numeric_digits       no

In [366]:
status(data)

variable  q_nan     p_nan  q_zeros   p_zeros  \
0                         fraudulent    455  0.730337      107  0.171750   
1   paymentMethodRegistrationFailure    281  0.451043      302  0.484751   
2                  paymentMethodType    281  0.451043        0  0.000000   
3              paymentMethodProvider    281  0.451043        0  0.000000   
4                paymentMethodIssuer    281  0.451043        0  0.000000   
5                  transactionAmount      0  0.000000        0  0.000000   
6                  transactionFailed      0  0.000000      455  0.730337   
7                        orderAmount    145  0.232745        0  0.000000   
8                         orderState    145  0.232745        0  0.000000   
9                        emailDomain      0  0.000000        0  0.000000   
10                     emailProvider      0  0.000000        0  0.000000   
11       customerIPAddressSimplified      0  0.000000        0  0.000000   
12                          sameCity      0  0.000000        0  0.000000   

    unique     type  
0        2   object  
1        2   object  
2        4   object  
3       10   object  
4       20   object  
5       67    int64  
6        2     bool  
7       67  float64  
8        3   object  
9        6   object  
10       5   object  
11       2   object  
12       3   object

## 5 - Guardar como csv

In [367]:
filename = "../datasets/converted_data/customer_dataset.csv"

data.to_csv(filename, index = False)